syntax = "proto2";

// The master service definition.
service MasterService {
  rpc SendMapperData(MapDataRequest) returns (MapDataResponse) {}
  rpc InvokeReducers(ReduceDataRequest) returns (ReduceDataResponse) {}
}

// The mapper service definition.
service MapperService {
  rpc SendDataPoint(DataPointRequest) returns (DataPointResponse) {}
}

// The reducer service definition.
service ReducerService {
  rpc SendPartition(PartitionRequest) returns (PartitionResponse) {}
}

message MapDataRequest {
  required int32 mapper_id = 1;
  required string ip = 6;
}

message MapDataResponse {
  repeated int32 input_split = 1;
  repeated Point centroids = 2;
  required string input_path = 3;
}

message Point {
  required float x = 1;
  required float y = 2;
}
message ReduceDataRequest {
  optional int32 reducers = 1;
  required string ip = 6;
}
message ReduceDataResponse {
  optional string message = 1;
}

// The message containing mapper data.
message MapperDataRequest {
  required int32 mapper_id = 1;
  required string ip = 6;
}

message MapperDataResponse {
  repeated float data_points = 1;
}

// The message containing a data point.
message DataPointRequest {
  required float data_point = 1;
  required string ip = 6;
}

message DataPointResponse {
  required int32 centroid_index = 1;
}

// The message containing a partition.
message PartitionRequest {
  required int32 reducer_id = 1;
  repeated KeyValue pairs = 2;
  required string ip = 6;
}

message PartitionResponse {
  optional string message = 1;
}

// A key-value pair.
message KeyValue {
  required float key = 1;
  repeated float value = 2;
}
